### PR TITLE
#3 k6 tests added for the new feature

### PR DESCRIPTION
**Pull request by codemascot**
#3 k6 tests added for the new feature

## Description
#3 k6 tests added for the new feature

## Related Issue(s)
- #3 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

### CODE REVIEW

File: test/k6/src/tests/instances.js [hunk 1]
```diff
@@ -291,6 +297,50 @@ function TC11_HardDeleteInstance(data) {
...
+  const instanceIdSplit = firstInstance["id"].split("/");
+
+  success = check(res, {
+    "TC12_GetInstances_ByPersonNumber: Get instance for party. InstanceId has expected format":
+      instanceIdSplit.length === 2,
+  });
+  addErrorCount(success);
+}
+
+//TC13 - Get all instances for party looked up with a person number
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:322]:
```suggestion
//TC13 - Get all instances for party looked up with an organisation number
```

File: test/k6/src/tests/instances.js [hunk 2]
```diff
@@ -291,6 +297,50 @@ function TC11_HardDeleteInstance(data) {
...
+    "TC13_GetInstances_ByOrgNumber: Get instance for party. Status is 200": (
+      r
+    ) => r.status === 200,
+  });
+  addErrorCount(success);
+
+  const firstInstance = JSON.parse(res.body).instances[0];
+  const instanceIdSplit = firstInstance["id"].split("/");
+
+  success = check(res, {
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:337]: A more interesting check here might be to check instanceOwner.organisationNumber to see if this matches the org number used as the instance owner identifier.

**Comment by acn-sbuad** [comment]
Updates to workflows to pass new environment variables is missing.

File: test/k6/src/tests/instances.js [hunk 3]
```diff
@@ -5,6 +5,7 @@
     -e env=*** `
     -e userId=*** `
     -e partyId=*** `
+    -e orgNumber=*** `
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:8]: I would add org number below existing user params. As to not create confusion of who the user and partyIds are linked to.

File: .github/workflows/regression-test-ATX.yml [hunk 4]
```diff
@@ -40,7 +40,7 @@ jobs:
       uses: grafana/k6-action@v0.3.1
       with:
         filename: test/k6/src/tests/instances.js
-        flags: -e runFullTestSet=true -e tokenGeneratorUserName=${{ secrets.TOKENGENERATOR_USERNAME }} -e tokenGeneratorUserPwd=${{ secrets.TOKENGENERATOR_USERPASSWORD }} -e env=${{ vars.ENV }} -e pid=${{ secrets.PERSON_NUMBER }} -e partyId=${{ secrets.PARTY_ID }} -e userId=${{ secrets.USER_ID }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e useTestTokenGenerator=true
+        flags: -e runFullTestSet=true -e tokenGeneratorUserName=${{ secrets.TOKENGENERATOR_USERNAME }} -e tokenGeneratorUserPwd=${{ secrets.TOKENGENERATOR_USERPASSWORD }} -e env=${{ vars.ENV }} -e pid=${{ secrets.PERSON_NUMBER }} -e partyId=${{ secrets.PARTY_ID }} -e userId=${{ secrets.USER_ID }} -e orgNumber=${{ secrets.ORG_NUMBER }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e useTestTokenGenerator=true
```
Comment (acn-sbuad) [inline .github/workflows/regression-test-ATX.yml:43]: Use case tests workkflow will also need updating with the same parameters :)

Comment (acn-sbuad) [reply .github/workflows/regression-test-ATX.yml:43]: TT and prod are still missing here. Reopening

File: test/k6/src/api/instances.js [hunk 5]
```diff
@@ -42,6 +42,14 @@ export function getInstances(token, filters) {
   return http.get(endpoint, params);
 }
 
+//Api call to Storage:Instances to get an instance by person number and return response
```
[the remark this answers is not in the document]
Comment (codemascot) [reply test/k6/src/api/instances.js:45]: Is there a specific rationale for this approach? 🤔 My reasoning for creating a separate function is to enhance maintainability and adaptability for new test cases based on the feature/purpose, without hampering the flow of already existing ones. And on the other hand, isn't adding new parameters makes the code more complex? Considering these scenarios, blending this to existing function do we achieve any particular benefit? 🤔

File: .github/workflows/regression-test-PROD.yml [hunk 6]
```diff
@@ -21,7 +21,7 @@ jobs:
       uses: grafana/k6-action@v0.3.1
       with:
         filename: test/k6/src/tests/instances.js
-        flags: -e runFullTestSet=true -e env=${{ vars.ENV }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e userName=${{ secrets.REGRESSION_USER_NAME}} -e userPassword=${{ secrets.REGRESSION_USER_PASSWORD}} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e mpClientId=${{ secrets.MP_CLIENT_ID }} -e mpKid=${{ secrets.MP_KID }} -e encodedJwk=${{ secrets.MP_ENCODEDJWK }}
+        flags: -e runFullTestSet=true -e env=${{ vars.ENV }} -e pid=${{ secrets.PERSON_NUMBER }} -e partyId=${{ secrets.PARTY_ID }} -e orgNumber=${{ secrets.ORG_NUMBER }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e userName=${{ secrets.REGRESSION_USER_NAME}} -e userPassword=${{ secrets.REGRESSION_USER_PASSWORD}} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e mpClientId=${{ secrets.MP_CLIENT_ID }} -e mpKid=${{ secrets.MP_KID }} -e encodedJwk=${{ secrets.MP_ENCODEDJWK }}
```
Comment (acn-sbuad) [inline .github/workflows/regression-test-PROD.yml:24]: why is partyId required here?

File: .github/workflows/regression-test-TT02.yml [hunk 7]
```diff
@@ -21,7 +21,7 @@ jobs:
       uses: grafana/k6-action@v0.3.1
       with:
         filename: test/k6/src/tests/instances.js
-        flags: -e runFullTestSet=true -e env=${{ vars.ENV }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e userName=${{ secrets.USE_CASE_USER_NAME}} -e userPassword=${{ secrets.USER_PASSWORD}} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e mpClientId=${{ secrets.MP_CLIENT_ID }} -e mpKid=${{ secrets.MP_KID }} -e encodedJwk=${{ secrets.MP_ENCODEDJWK }}
+        flags: -e runFullTestSet=true -e env=${{ vars.ENV }} -e pid=${{ secrets.PERSON_NUMBER }} -e partyId=${{ secrets.PARTY_ID }} -e orgNumber=${{ secrets.ORG_NUMBER }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e userName=${{ secrets.USE_CASE_USER_NAME}} -e userPassword=${{ secrets.USER_PASSWORD}} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e mpClientId=${{ secrets.MP_CLIENT_ID }} -e mpKid=${{ secrets.MP_KID }} -e encodedJwk=${{ secrets.MP_ENCODEDJWK }}
```
Comment (acn-sbuad) [inline .github/workflows/regression-test-TT02.yml:24]: Same goes here. Why is partyId required?

File: test/k6/src/tests/instances.js [hunk 8]
```diff
@@ -291,6 +295,53 @@ function TC11_HardDeleteInstance(data) {
   addErrorCount(success);
 }
 
+//TC12 - Get all instances for party looked up with a person number
+function TC12_GetInstances_ByPersonNumber(data) {
+  var instanceOwnerIdentifier = "Person:" + data.personNumber;
```
[the remark this answers is not in the document]
Comment (acn-sbuad) [reply test/k6/src/tests/instances.js:300]: Unresolving. Adding a condition only means if we don't get any instances returned we don't check for the existence of instances either.. In it's current state this test does in no way verify that the lookup functionality is working as expected. 

If no instance is returned in the list the test should fail.

File: .github/workflows/regression-test-PROD.yml (hunk 6 above)
Comment (codemascot) [reply .github/workflows/regression-test-PROD.yml:24]: Yea! Correct! We don't need party ID when using basic auth! 👍

Comment (acn-sbuad) [reply .github/workflows/regression-test-PROD.yml:24]: so the property should be removed. Unresolving

File: test/k6/src/tests/instances.js [hunk 9]
```diff
@@ -276,15 +280,83 @@ function TC10_SoftDeleteInstance(data) {
...
+
+    success = check(res, {
+      "TC11_GetInstances_ByPersonNumber: Get instance for party. InstanceId has expected format":
+        instanceIdSplit.length === 2,
+    });
+    addErrorCount(success);
+  }
+}
+
+//TC13 - Get all instances for party looked up with an organisation number
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:314]: mismatch tc12 and t13

File: test/k6/src/tests/instances.js [hunk 10]
```diff
@@ -276,15 +280,83 @@ function TC10_SoftDeleteInstance(data) {
...
+function TC12_GetInstances_ByOrgNumber(data) {
+  var instanceOwnerIdentifier = "Organisation:" + data.orgNumber;
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:316]: are we ensuring before this tests runs that there actually exists any instances for the given orgnumber?

File: test/k6/src/tests/instances.js [hunk 11]
```diff
@@ -276,15 +280,83 @@ function TC10_SoftDeleteInstance(data) {
...
+  var success = check(res, {
+    "TC12_GetInstances_ByOrgNumber: Get instance for party. Status is 200": (
+      r
+    ) => r.status === 200,
+  });
+  addErrorCount(success);
+
+  const firstInstance = JSON.parse(res.body).instances[0];
+  if (firstInstance) { // Making sure there actually exists instances for the instanceOwner
+    const orgnasationNumber = firstInstance.instanceOwner.organisationNumber;
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:334]: typo `orgnasationNumber`

Comment (codemascot) [reply test/k6/src/tests/instances.js]: This is resolved!

File: .github/workflows/regression-test-PROD.yml (hunk 6 above)
Comment (codemascot) [reply .github/workflows/regression-test-PROD.yml]: My bad! Double checked this time! `-e partyId=${{ secrets.PARTY_ID }}` removed from all the 4 files. 👍

File: .github/workflows/regression-test-TT02.yml (hunk 7 above)
Comment (codemascot) [reply .github/workflows/regression-test-TT02.yml]: https://github.com/Altinn/altinn-storage/pull/381#discussion_r1611101856

File: test/k6/src/tests/instances.js [hunk 12]
```diff
@@ -276,15 +280,83 @@ function TC10_SoftDeleteInstance(data) {
   addErrorCount(success);
 }
 
-function TC11_HardDeleteInstance(data) {
+//TC12 - Get all instances for party looked up with a person number
```
[the remark this answers is not in the document]
Comment (codemascot) [reply test/k6/src/tests/instances.js]: Checked! Now this has been fixed! Marking this resolved! 👍

File: test/k6/src/tests/instances.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (codemascot) [reply test/k6/src/tests/instances.js:285]: Checked! 👍 Updated with a mechanism to test the existence of the instances.

File: test/k6/src/tests/instances.js [hunk 13]
```diff
@@ -73,12 +77,17 @@ export function setup() {
 
 // TC01 - POST instance
 function TC01_PostInstance(data) {
+  const options = {
+    "personNumber": data.personNumber
+  };
+
   var res = instancesApi.postInstance(
-    data.userToken,
+    data.orgToken,
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:85]: no need to change this to org token. The user is authorized to instantiate on behalf of themselves.

File: test/k6/src/tests/instances.js [hunk 14]
```diff
@@ -73,12 +77,17 @@ export function setup() {
 
 // TC01 - POST instance
 function TC01_PostInstance(data) {
+  const options = {
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:80]: Would reconsider including the options here. Look into how partyId and appId is set for the instance before sending to storage. Could personNumber be handled the same way? Or should we rather include all the properties in options? Using both approaches seem a bit messy.

File: test/k6/src/tests/instances.js [hunk 15]
```diff
@@ -276,15 +285,134 @@ function TC10_SoftDeleteInstance(data) {
...
+  addErrorCount(instanciationResultSuccess);
+
+  if (dataBody.count > 0) { // If instance(s) exists
+    const firstInstance = dataBody.instances[0];
+    const personNumber = firstInstance.instanceOwner.personNumber;
+    const instanceIdSplit = firstInstance["id"].split("/");
+
+    const success = check(res, {
+      "TC11_GetInstances_ByPersonNumber: Get instance for party. InstanceId has expected format":
+        instanceIdSplit.length === 2,
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:317]: once we get to this test we are already confident the format is as expected as this is covered by an earlier TC.  Can limit the number of checks to what is relevant for the feature being tested.

File: test/k6/src/tests/instances.js (hunk 13 above)
Comment (codemascot) [reply test/k6/src/tests/instances.js:85]: Correct, as this is for the person number test.

File: test/k6/src/tests/instances.js (hunk 14 above)
Comment (codemascot) [reply test/k6/src/tests/instances.js:80]: I would go for putting all the properties in options as it seems easier to handle validating specific parameters. Plus, option represents both `personNumber` and `orgNumber` which seemed a bit easy to handle than passing two parameters when calling this function. Hereby, my money is on passing the parameters as an object rather separate separate parameters. 🤔

File: test/k6/src/tests/instances.js [hunk 16]
```diff
@@ -276,15 +286,130 @@ function TC10_SoftDeleteInstance(data) {
...
+    ) => r.status === 200,
+    "TC11_GetInstances_ByPersonNumber: Instances exists for party.": 
+    dataBody.count > 0,
+  });
+  addErrorCount(instanciationResultSuccess);
+
+  if (dataBody.count > 0) { // If instance(s) exists
+    const firstInstance = dataBody.instances[0];
+    const personNumber = firstInstance.instanceOwner.personNumber;
+    const instanceIdSplit = firstInstance["id"].split("/");
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:314]: constant is not used anywhere

File: test/k6/src/tests/instances.js [hunk 17]
```diff
@@ -276,15 +286,129 @@ function TC10_SoftDeleteInstance(data) {
...
+    const successHardDelete = check(hardDeletionResult, {
+      "TC12_GetInstances_ByOrgNumber: Hard delete instance for organisation. Status is 200": (r) =>
+        r.status === 200,
+    });
+  
+    addErrorCount(successHardDelete);
+  }
+}
+
+// Here we are deleting instances,
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:399]: no need for this comment

File: .github/workflows/use-case-TT02.yml [hunk 18]
```diff
@@ -22,7 +22,7 @@ jobs:
       uses: grafana/k6-action@v0.3.1
       with:
         filename: test/k6/src/tests/instances.js
-        flags: -e env=${{ vars.ENV }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e userName=${{ secrets.USE_CASE_USER_NAME}} -e userPassword=${{ secrets.USER_PASSWORD}} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e mpClientId=${{ secrets.MP_CLIENT_ID }} -e mpKid=${{ secrets.MP_KID }} -e encodedJwk=${{ secrets.MP_ENCODEDJWK }}
+        flags: -e env=${{ vars.ENV }} -e pid=${{ secrets.PERSON_NUMBER }} -e orgNumber=${{ secrets.ORG_NUMBER }} -e orgPartyId=${{ secrets.ORG_PARTY_ID }} -e org=${{ vars.ORG }} -e app=${{ vars.APP }} -e userName=${{ secrets.USE_CASE_USER_NAME}} -e userPassword=${{ secrets.USER_PASSWORD}} -e apimSubsKey=${{ secrets.APIM_SUBSKEY }} -e mpClientId=${{ secrets.MP_CLIENT_ID }} -e mpKid=${{ secrets.MP_KID }} -e encodedJwk=${{ secrets.MP_ENCODEDJWK }}
```
Comment (acn-sbuad) [inline .github/workflows/use-case-TT02.yml:25]: I don't think we need to add the org tests to the use case tests. Would suffice to have them in the regression test and only include one of the new tests in the use case tests

File: test/k6/src/tests/instances.js [hunk 19]
```diff
@@ -9,6 +9,8 @@
     -e username=*** `
     -e userpwd=*** `
     -e org=ttd `
+    -e orgNumber=*** `
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:12]: let's not split up org, app which are naturally grouped together here. OrgNumber and orgPartyId do not represent the same entity as org. New properties could be placed before org instead?

File: .github/workflows/use-case-TT02.yml (hunk 18 above)
Comment (acn-sbuad) [reply .github/workflows/use-case-TT02.yml]: you've removed the passing of environment variables, but you will also need to update the test file so test case isn't run outside of the full test set.

File: test/k6/src/api/instances.js [hunk 20]
```diff
@@ -5,40 +5,52 @@ import { stopIterationOnFail } from "../errorhandler.js";
 
 /**
  * Api call to Storage:Instances to create an app instance and returns response
- * @param {*} token token value to be sent in apiHelper for authentication
- * @param {*} partyId party id of the user to whom instance is to be created
- * @param {*} org short name of ap owner
- * @param {*} app app name
- * @param {JSON} serializedInstance instance json metadata sent in request body
- * @returns {JSON} Json object including response apiHelperss, body, timings
+ * @param {Object} instanciationOptions Object containing the parameters for the API call.
```
Comment (acn-sbuad) [inline test/k6/src/api/instances.js:8]: spelling error `instanciation` should be `instantiation`

Comment (acn-sbuad) [inline test/k6/src/api/instances.js:8]: spelling error `instanciation` should be `instantiation`

Comment (codemascot) [reply test/k6/src/api/instances.js]: What a silly mistake! I should have been more careful! My bad! Fixed now! 👍 Marking is resolved. 🙂

File: test/k6/src/tests/instances.js [hunk 21]
```diff
@@ -276,15 +286,125 @@ function TC10_SoftDeleteInstance(data) {
...
+    });
+    addErrorCount(hasInstances);
+  
+    if (dataBody.count > 0) { // If instance(s) exists
+      const firstInstance = dataBody.instances[0];
+      const orgNumber = firstInstance.instanceOwner.organisationNumber;
+      const instanceIdSplit = firstInstance["id"].split("/");
+  
+      const successResponse = check(res, {
+        "TC12_GetInstances_ByOrgNumber: Get instance for party. InstanceId has expected format":
```
Comment (acn-sbuad) [inline test/k6/src/tests/instances.js:375]: This check has already been executed in a previous test. No need to include here as well

File: test/k6/src/archive/rf0002.js [hunk 22]
```diff
@@ -51,8 +51,16 @@ export default function () {
   const partyId = data['orgPartyId'];
   instanceId = '';
 
-  //Test to create an instance with storage api and validate the response
-  res = instances.postInstance(runtimeToken, partyId, appOwner, level2App, instanceJson);
+  //Test to create an instance with storage api and validate the response // token, partyId, org, app, serializedInstance
```
Comment (acn-sbuad) [inline test/k6/src/archive/rf0002.js:54]:
```suggestion
  //Test to create an instance with storage api and validate the response 
```

**Review by acn-sbuad — APPROVED** [review]
🎈

**Pull request merged**